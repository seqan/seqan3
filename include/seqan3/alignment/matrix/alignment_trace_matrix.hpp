// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------

/*!\file
 * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
 * \brief Contains the declaration of seqan3::detail::alignment_trace_matrix.
 */

#pragma once

#include <seqan3/alignment/matrix/alignment_score_matrix.hpp>
#include <seqan3/alignment/matrix/matrix_concept.hpp>
#include <seqan3/alignment/matrix/trace_directions.hpp>

namespace seqan3::detail
{

//!\brief The declaration of alignment_trace_matrix. Each definition of this
//!       declaration must satisfy seqan3::detail::Matrix.
//!\ingroup alignment_matrix
//!\implements seqan3::detail::Matrix
template <typename ...>
class alignment_trace_matrix;

/*!\brief A trace matrix represented in a one-dimensional std::vector
 * \ingroup alignment_matrix
 *
 * \details
 *
 * This data structure stores the matrix in a flat way using the
 * std::vector<#entry_type> data structure where each row is stored
 * continuously.
 *
 * # Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_trace_matrix_vector.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_trace_matrix_vector.out
 */
template <>
class alignment_trace_matrix<std::vector<trace_directions>>
    : public row_wise_matrix<trace_directions>
{
public:
    using row_wise_matrix<trace_directions>::row_wise_matrix;
};

/*!\brief A trace matrix that uses an underlying seqan3::detail::alignment_score_matrix
 * \ingroup alignment_matrix
 * \tparam database_type            The type of the database sequence.
 * \tparam query_type               The type of the query sequence.
 * \tparam align_config_type        The type of the alignment config.
 * \tparam ...score_matrix_params_t The template parameters of seqan3::detail::alignment_score_matrix
 *
 * \details
 *
 * This data structure uses directly the score matrix to infer the trace matrix
 * and works for any seqan3::detail::alignment_score_matrix.
 *
 * \todo TODO: Is currently only able to handle the edit distance.
 *
 * # Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_trace_matrix.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_trace_matrix.out
 */
template <typename database_type, typename query_type, typename align_config_type, typename ...score_matrix_params_t>
//!\cond
    requires Matrix<alignment_score_matrix<score_matrix_params_t...>> &&
             std::Integral<typename alignment_score_matrix<score_matrix_params_t...>::entry_type>
//!\endcond
class alignment_trace_matrix<database_type, query_type, align_config_type, alignment_score_matrix<score_matrix_params_t...>>
    : public alignment_score_matrix<score_matrix_params_t...>
{
public:
    //!\brief The type of the score matrix.
    using score_matrix_type = alignment_score_matrix<score_matrix_params_t...>;

    //!\brief The type of an entry in the score matrix.
    using score_type = typename score_matrix_type::entry_type;

    //!\copydoc seqan3::detail::Matrix::entry_type
    using entry_type = trace_directions;

    /*!\name Constructors, destructor and assignment
     * \{
     */
     alignment_trace_matrix() = default;                                           //!< Defaulted
     alignment_trace_matrix(alignment_trace_matrix const &) = default;             //!< Defaulted
     alignment_trace_matrix(alignment_trace_matrix &&) = default;                  //!< Defaulted
     alignment_trace_matrix & operator=(alignment_trace_matrix const &) = default; //!< Defaulted
     alignment_trace_matrix & operator=(alignment_trace_matrix &&) = default;      //!< Defaulted
     ~alignment_trace_matrix() = default;                                          //!< Defaulted

    /*!\brief Construct the trace matrix by using a score_matrix.
     * \param database     The database sequence.
     * \param query        The query sequence.
     * \param config       The alignment config.
     * \param score_matrix The score matrix.
     */
    alignment_trace_matrix(database_type database, query_type query, align_config_type config, score_matrix_type score_matrix)
        : score_matrix_type(std::move(score_matrix)),
          _database{std::forward<database_type>(database)},
          _query{std::forward<query_type>(query)},
          _config{std::forward<align_config_type>(config)}
    {}
    //!\}

    //!\copydoc seqan3::detail::Matrix::rows
    using score_matrix_type::rows;
    //!\copydoc seqan3::detail::Matrix::cols
    using score_matrix_type::cols;

    //!\brief The trace directions of the matrix at position (*row*, *col*).
    entry_type at(size_t const row, size_t const col) const noexcept
    {
        entry_type direction{};

        if (is_trace_diagonal(row, col))
            direction |= entry_type::diagonal;

        if (is_trace_up(row, col))
            direction |= entry_type::up;

        if (is_trace_left(row, col))
            direction |= entry_type::left;

        return direction;
    }

    //!\brief Access to the score_matrix.
    score_matrix_type const & score_matrix() const noexcept
    {
        return *this;
    }

private:

    //!\brief Does the trace come from the above entry?
    bool is_trace_up(size_t const row, size_t const col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type gap = -1;

        score_type curr = score_matrix().at(row, col);
        score_type up = row == 0 ? -col : score_matrix().at(row - 1, col);
        return curr == up + gap;
    }

    //!\brief Does the trace come from the left entry?
    bool is_trace_left(size_t const row, size_t const col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type gap = -1;

        score_type curr = score_matrix().at(row, col);
        score_type left = col == 0 ? -row : score_matrix().at(row, col - 1);
        return curr == left + gap;
    }

    //!\brief Does the trace come from the diagonal entry?
    bool is_trace_diagonal(size_t const row, size_t const col) const noexcept
    {
        // TODO: use the alignment_config to calculate the score
        score_type match = 0;
        score_type mismatch = -1;

        score_type curr = score_matrix().at(row, col);
        if (col == 0 || row == 0)
            return false;

        score_type diag = score_matrix().at(row - 1, col - 1);
        bool is_match = _query[row - 1] == _database[col - 1];

        return (is_match && curr == diag + match) ||
              (!is_match && curr == diag + mismatch);
    }

    //!\brief The database sequence.
    database_type _database;
    //!\brief The query sequence.
    query_type _query;
    //!\brief The alignment config.
    align_config_type _config;
};

/*!\name Type deduction guides
 * \relates seqan3::detail::alignment_trace_matrix
 * \{
 */

//!\brief Deduce the trace matrix type from the provided arguments.
alignment_trace_matrix(std::vector<trace_directions>, size_t rows, size_t cols)
    -> alignment_trace_matrix<std::vector<trace_directions>>;

//!\brief Deduce the trace matrix type from the provided arguments.
template <typename database_t, typename query_t, typename align_config_t, typename alignment_t, typename ...options_t>
alignment_trace_matrix(database_t && database, query_t && query, align_config_t && config, alignment_score_matrix<alignment_t, options_t...>)
    -> alignment_trace_matrix<database_t, query_t, align_config_t, alignment_score_matrix<alignment_t, options_t...>>;
//!\}

} // namespace seqan3::detail
