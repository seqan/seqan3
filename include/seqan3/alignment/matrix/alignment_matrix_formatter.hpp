// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------

/*!\file
 * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
 * \brief Contains seqan3::detail::alignment_matrix_formatter and seqan3::detail::alignment_matrix_format.
 */

#pragma once

#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>

#include <seqan3/alignment/matrix/matrix_concept.hpp>
#include <seqan3/alignment/matrix/alignment_trace_matrix.hpp>
#include <seqan3/core/debug_stream.hpp>

namespace seqan3::detail
{

/*!\brief Format used by seqan3::detail::alignment_matrix_formatter
 * \ingroup alignment_matrix
 *
 * \details
 *
 * With seqan3::detail::alignment_matrix_format you can style:
 *  * the epsilon symbol (alignment_matrix_format::epsilon)
 *  * the column symbol that separates each cell in a row (alignment_matrix_format::col_sep)
 *  * the row symbol that divides each row (alignment_matrix_format::row_sep)
 *  * the column symbol that comes after the row symbol (alignment_matrix_format::row_col_sep)
 *  * the traces symbols of a traceback matrix (alignment_matrix_format::trace_dir)
 *
 * # Example
 * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.cpp alignment_matrix_format
 *
 * ### Output
 * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::out
 */
struct alignment_matrix_format
{
    //!\brief epsilon symbol (a single symbol)
    char const * epsilon{};
    //!\brief column separator symbol (a single symbol)
    char const * col_sep{};
    //!\brief row separator symbol (a single symbol)
    char const * row_sep{};
    //!\brief row separator symbol (a single symbol)
    char const * row_col_sep{};
    //!\brief infinity symbol (a single symbol)
    char const * inf{};

    /*!\brief Eight symbols for each combination of directions a trace can have (each entry can have multiple symbols).
     *
     * * 1st bit: **D** = diagonal
     * * 2nd bit: **U** = up
     * * 3th bit: **L** = left
     *
     * | i     | 1st bit | 2nd bit | 3rd bit | trace_dir[i] |
     * | :---: |---------|---------|---------| :----------: |
     * | **0** | 0       | 0       | 0       | **No dir**   |
     * | **1** | 1       | 0       | 0       | **D**        |
     * | **2** | 0       | 1       | 0       | **U**        |
     * | **3** | 1       | 1       | 0       | **DU**       |
     * | **4** | 0       | 0       | 1       | **L**        |
     * | **5** | 1       | 0       | 1       | **DL**       |
     * | **6** | 0       | 1       | 1       | **UL**       |
     * | **7** | 1       | 1       | 1       | **DUL**      |
     *
     */
    char const * trace_dir[8]{};

    /*!\brief The CSV format that makes it easy to export the matrix
     * \sa https://en.wikipedia.org/wiki/Comma-separated_values
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::csv#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::csv#trace
     */
    static const alignment_matrix_format csv;

    /*!\brief A format that uses only ascii symbols.
     * \sa https://en.wikipedia.org/wiki/Ascii
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::ascii#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::ascii#trace
     */
    static const alignment_matrix_format ascii;

    /*!\brief A format that uses unicode block symbols.
     * \sa https://en.wikipedia.org/wiki/Unicode
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_block#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_block#trace
     */
    static const alignment_matrix_format unicode_block;

    /*!\brief A format that uses unicode braille symbols.
     * \sa https://en.wikipedia.org/wiki/Unicode
     *
     *
     * ### Score output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_braille#score
     *
     * ### Trace output
     * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_braille#trace
     */
    static const alignment_matrix_format unicode_braille;

   /*!\brief A format that uses unicode arrow symbols.
    * \sa https://en.wikipedia.org/wiki/Unicode
    *
    *
    * ### Score output
    * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_arrows#score
    *
    * ### Trace output
    * \snippet test/snippet/alignment/matrix/alignment_matrix_formatter.out alignment_matrix_format::unicode_arrows#trace
    */
    static const alignment_matrix_format unicode_arrows;
};

constexpr alignment_matrix_format alignment_matrix_format::csv
{
    " ", ";", "", "", "",
    {"N","D","U","DU","L","DL","UL","DUL"}
};

constexpr alignment_matrix_format alignment_matrix_format::ascii
{
    " ", "|", "-", "/", "INF",
    {" ","D","U","DU","L","DL","UL","DUL"}
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_block
{
    "ε", "║", "═", "╬", "∞",
    {"█","▘","▝","▀","▖","▌","▞","▛"}
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_braille
{
    "ε", "║", "═", "╬", "∞",
    {"⠀","⠁","⠈","⠉","⠄","⠅","⠌","⠍"}
};

constexpr alignment_matrix_format alignment_matrix_format::unicode_arrows
{
    "ε", "║", "═", "╬", "∞",
    {"↺","↖","↑","↖↑","←","↖←","↑←","↖↑←"}
};

/*!\brief Formats and prints trace and score matrices that satisfy the
 *        seqan3::detail::Matrix.
 * \ingroup alignment_matrix
 * \tparam alignment_matrix_t The matrix class which satisfies the seqan3::detail::Matrix
 *
 * \details
 *
 * # Example
 *
 * \snippet test/snippet/alignment/matrix/alignment_score_matrix.cpp code
 *
 * ### Output
 * \include test/snippet/alignment/matrix/alignment_score_matrix.out
 */
template <Matrix alignment_matrix_t>
class alignment_matrix_formatter
{
public:
    //!\brief The type of the #matrix
    using alignment_matrix_type = alignment_matrix_t;

private:
    //!\brief The matrix to format.
    alignment_matrix_type const & matrix;

public:
    //!\brief The actual format used by the formatter.
    alignment_matrix_format symbols;

    /*!\name Constructors, destructor and assignment
     * \{
     */
     alignment_matrix_formatter() = delete;                                                //!< Matrix taken as ref.
     alignment_matrix_formatter(alignment_matrix_formatter const &) = default;             //!< Defaulted
     alignment_matrix_formatter(alignment_matrix_formatter &&) = default;                  //!< Defaulted
     alignment_matrix_formatter & operator=(alignment_matrix_formatter const &) = default; //!< Defaulted
     alignment_matrix_formatter & operator=(alignment_matrix_formatter &&) = default;      //!< Defaulted
     ~alignment_matrix_formatter() = default;                                              //!< Defaulted

    /*!\brief Construct the matrix out of the *entries*, the *rows*,
     *        and the *cols*.
     * \param _matrix  \copydoc matrix
     * \param _symbols \copydoc symbols
     */
    alignment_matrix_formatter(alignment_matrix_type const & _matrix, alignment_matrix_format _symbols = alignment_matrix_format::unicode_arrows)
        : matrix{_matrix}, symbols{_symbols}
    {}
    //!\}

    //!\brief \copydoc seqan3::detail::Matrix::entry_type
    using entry_type = typename alignment_matrix_type::entry_type;

    //!\brief Whether #alignment_matrix_type is a traceback matrix.
    static constexpr bool is_traceback_matrix = std::is_same_v<entry_type, trace_directions>;

    //!\brief Determines the largest width of all entries in the #matrix,
    //!       e.g. `-152` has width 4.
    size_t auto_width() const noexcept
    {
        size_t col_width = 1;
        for (size_t row = 0; row < matrix.rows(); ++row)
            for (size_t col = 0; col < matrix.cols(); ++col)
                col_width = std::max(col_width, unicode_str_length(entry_at(row, col)));
        return col_width;
    }

    //!\brief Print the formatted #matrix to std::cout.
    //!\param[in]  database        the database sequence
    //!\param[in]  query           the query sequence
    //!\param[in]  column_width    width of each cell, std::nullopt defaults to auto_width()
    template <typename database_t, typename query_t>
    void format(database_t && database, query_t && query, std::optional<size_t> column_width = std::nullopt) const noexcept
    {
        format(std::forward<database_t>(database), std::forward<query_t>(query), std::cout, column_width);
    }

    /*!\brief Print the formatted #matrix to the *cout* stream.
     * \tparam         char_t          char type of std::basic_ostream
     * \tparam         traits_t        traits type of std::basic_ostream
     * \param[in]      database        the database sequence
     * \param[in]      query           the query sequence
     * \param[in,out]  cout            print formatted #matrix into this ostream
     * \param[in]      column_width    width of each cell, std::nullopt defaults to auto_width()
     */
    template <typename database_t, typename query_t, typename oostream_t>
    void format(database_t && database, query_t && query, oostream_t & cout, std::optional<size_t> const column_width) const noexcept
    {
        size_t const _column_width = column_width.has_value() ? column_width.value() : auto_width();

        auto print_cell = [&](std::string const & symbol)
        {
            // deal with unicode chars that mess up std::setw
            size_t const length_bytes = unicode_str_length_bytes(symbol);
            size_t const length = unicode_str_length(symbol);
            size_t const offset = length_bytes - length;

            cout << std::left
                 << std::setw(_column_width + offset)
                 << symbol
                 << symbols.col_sep;
        };

        auto print_first_cell = [&](std::string const & symbol)
        {
            cout << symbol << symbols.col_sep;
        };

        // |_|d|a|t|a|b|a|s|e|
        auto print_first_row = [&]
        {
            print_first_cell(" ");
            print_cell(symbols.epsilon);

            for (size_t col = 0; col < matrix.cols() - 1; ++col)
                print_cell(as_string(database[col]));
            cout << "\n";
        };

        // |-|-|-|-|-|-|-|-|-|
        auto print_divider = [&]
        {
            cout << " " << symbols.row_col_sep;
            for (size_t col = 0; col < matrix.cols(); ++col)
            {
                for (size_t i = 0; i < _column_width; ++i)
                    cout << symbols.row_sep;
                cout << symbols.row_col_sep;
            }
            cout << "\n";
        };

        print_first_row();
        for (size_t row = 0; row < matrix.rows(); ++row)
        {
            if (symbols.row_sep[0] != '\0')
                print_divider();

            // one query letter + one row of scores / traces
            if (row == 0)
                print_first_cell(symbols.epsilon);
            else
                print_first_cell(as_string(query[row - 1]));
            for (size_t col = 0; col < matrix.cols(); ++col)
                print_cell(entry_at(row, col));
            cout << "\n";
        }
    }

private:
    //!\brief Same as #matrix\.at(*row*, *col*), but converts the value to
    //!       a trace symbol (alignment_matrix_format::trace_dir) if the
    //!       #matrix is a traceback matrix.
    std::string entry_at(size_t const row, size_t const col) const noexcept
    {
        if constexpr(is_traceback_matrix)
        {
            trace_directions direction = matrix.at(row, col);
            return symbols.trace_dir[(size_t)(direction) % 8u];
        }
        else
        {
            entry_type entry = matrix.at(row, col);
            if (entry == matrix_inf<entry_type>)
                return as_string(symbols.inf);
            return as_string(entry);
        }
    }

    //!\brief Convert a matrix entry into a std::string
    template <typename entry_type>
    static std::string as_string(entry_type && entry) noexcept
    {
        std::stringstream strstream;
        debug_stream_type stream{strstream};
        stream << entry;
        return strstream.str();
    }

    //!\brief The length of the *str* (traceback symbols are unicode aware)
    //!\sa https://en.wikipedia.org/wiki/UTF-8 for encoding details
    static size_t unicode_str_length(std::string const & str) noexcept
    {
        size_t length = 0u;
        for (auto it = str.cbegin(), it_end = str.cend(); it < it_end; ++it, ++length)
        {
            uint8_t v = *it;
            if ((v & 0b11100000) == 0b11000000)
                ++it;
            else if ((v & 0b11110000) == 0b11100000)
                it += 2;
            else if ((v & 0b11111000) == 0b11110000)
                it += 3;
        }
        return length;
    }

    //!\brief The number of bytes the *str* uses
    static size_t unicode_str_length_bytes(std::string const & str) noexcept
    {
        return str.length();
    }

    //!\brief Befriend test case
    friend struct matrix_formatter_test;
};

/*!\name Type deduction guides
 * \relates seqan3::detail::alignment_matrix_formatter
 * \{
 */

//!\brief Deduce the matrix type from the provided argument.
template<typename alignment_matrix_t>
alignment_matrix_formatter(alignment_matrix_t const &) -> alignment_matrix_formatter<alignment_matrix_t>;

//!\brief Deduce the matrix type from the provided arguments.
template<typename alignment_matrix_t>
alignment_matrix_formatter(alignment_matrix_t const &, alignment_matrix_format) -> alignment_matrix_formatter<alignment_matrix_t>;
//!\}

} // namespace seqan3::detail
