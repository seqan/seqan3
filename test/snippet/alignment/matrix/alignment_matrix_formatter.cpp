#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alignment/matrix/alignment_matrix_formatter.hpp>
#include <seqan3/alignment/matrix/alignment_trace_matrix.hpp>
#include <seqan3/core/debug_stream.hpp>

int main()
{
    using namespace seqan3;
    using namespace seqan3::detail;

    struct no_config
    {};

    std::vector<dna4> database = "AACCGGTT"_dna4;
    std::vector<dna4> query = "ACGT"_dna4;

    alignment_score_matrix score_matrix
    {
        std::vector
        {
            0, 1, 2, 3, 4, 5, 6, 7, 8,
            1, 0, 1, 2, 3, 4, 5, 6, 7,
            2, 1, 1, 1, 2, 3, 4, 5, 6,
            3, 2, 2, 2, 2, 2, 3, 4, 5,
            4, 3, 3, 3, 3, 3, 3, 3, 4
        },
        5u,
        9u
    };

    alignment_trace_matrix trace_matrix{database, query, no_config{}, score_matrix};

    debug_stream << "//! [alignment_matrix_format::out]" << std::endl;
//! [alignment_matrix_format]
using namespace seqan3;

alignment_matrix_format format
{
    "e", // epsilon
    "|", // col_sep
    "-", // row_sep
    "/", // row_col_sep
    "INF", // infinity
    {" ","D","U","DU","L","DL","UL","DUL"} // trace_dir
};

alignment_matrix_formatter{trace_matrix, format}.format(database, query);
//! [alignment_matrix_format]
    debug_stream << "//! [alignment_matrix_format::out]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::ascii#score]" << std::endl;
    alignment_matrix_formatter{score_matrix, alignment_matrix_format::ascii}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::ascii#score]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::ascii#trace]" << std::endl;
    alignment_matrix_formatter{trace_matrix, alignment_matrix_format::ascii}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::ascii#trace]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::csv#score]" << std::endl;
    alignment_matrix_formatter{score_matrix, alignment_matrix_format::csv}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::csv#score]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::csv#trace]" << std::endl;
    alignment_matrix_formatter{trace_matrix, alignment_matrix_format::csv}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::csv#trace]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_block#score]" << std::endl;
    alignment_matrix_formatter{score_matrix, alignment_matrix_format::unicode_block}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_block#score]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_block#trace]" << std::endl;
    alignment_matrix_formatter{trace_matrix, alignment_matrix_format::unicode_block}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_block#trace]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_braille#score]" << std::endl;
    alignment_matrix_formatter{score_matrix, alignment_matrix_format::unicode_braille}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_braille#score]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_braille#trace]" << std::endl;
    alignment_matrix_formatter{trace_matrix, alignment_matrix_format::unicode_braille}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_braille#trace]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_arrows#score]" << std::endl;
    alignment_matrix_formatter{score_matrix, alignment_matrix_format::unicode_arrows}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_arrows#score]" << std::endl;

    debug_stream << "//! [alignment_matrix_format::unicode_arrows#trace]" << std::endl;
    alignment_matrix_formatter{trace_matrix, alignment_matrix_format::unicode_arrows}.format(database, query);
    debug_stream << "//! [alignment_matrix_format::unicode_arrows#trace]" << std::endl;

    return 0;
}
