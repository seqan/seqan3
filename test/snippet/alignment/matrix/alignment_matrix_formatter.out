//! [alignment_matrix_format::out]
 |e |A |A |C |C |G |G |T |T |
 /--/--/--/--/--/--/--/--/--/
e|  |L |L |L |L |L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
A|U |D |DL|L |L |L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
C|U |U |D |D |DL|L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
G|U |U |DU|DU|D |D |DL|L |L |
 /--/--/--/--/--/--/--/--/--/
T|U |U |DU|DU|DU|DU|D |D |DL|
//! [alignment_matrix_format::out]
//! [alignment_matrix_format::ascii#score]
 | |A|A|C|C|G|G|T|T|
 /-/-/-/-/-/-/-/-/-/
 |0|1|2|3|4|5|6|7|8|
 /-/-/-/-/-/-/-/-/-/
A|1|0|1|2|3|4|5|6|7|
 /-/-/-/-/-/-/-/-/-/
C|2|1|1|1|2|3|4|5|6|
 /-/-/-/-/-/-/-/-/-/
G|3|2|2|2|2|2|3|4|5|
 /-/-/-/-/-/-/-/-/-/
T|4|3|3|3|3|3|3|3|4|
//! [alignment_matrix_format::ascii#score]
//! [alignment_matrix_format::ascii#trace]
 |  |A |A |C |C |G |G |T |T |
 /--/--/--/--/--/--/--/--/--/
 |  |L |L |L |L |L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
A|U |D |DL|L |L |L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
C|U |U |D |D |DL|L |L |L |L |
 /--/--/--/--/--/--/--/--/--/
G|U |U |DU|DU|D |D |DL|L |L |
 /--/--/--/--/--/--/--/--/--/
T|U |U |DU|DU|DU|DU|D |D |DL|
//! [alignment_matrix_format::ascii#trace]
//! [alignment_matrix_format::csv#score]
 ; ;A;A;C;C;G;G;T;T;
 ;0;1;2;3;4;5;6;7;8;
A;1;0;1;2;3;4;5;6;7;
C;2;1;1;1;2;3;4;5;6;
G;3;2;2;2;2;2;3;4;5;
T;4;3;3;3;3;3;3;3;4;
//! [alignment_matrix_format::csv#score]
//! [alignment_matrix_format::csv#trace]
 ;  ;A ;A ;C ;C ;G ;G ;T ;T ;
 ;N ;L ;L ;L ;L ;L ;L ;L ;L ;
A;U ;D ;DL;L ;L ;L ;L ;L ;L ;
C;U ;U ;D ;D ;DL;L ;L ;L ;L ;
G;U ;U ;DU;DU;D ;D ;DL;L ;L ;
T;U ;U ;DU;DU;DU;DU;D ;D ;DL;
//! [alignment_matrix_format::csv#trace]
//! [alignment_matrix_format::unicode_block#score]
 ║ε║A║A║C║C║G║G║T║T║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
ε║0║1║2║3║4║5║6║7║8║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
A║1║0║1║2║3║4║5║6║7║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
C║2║1║1║1║2║3║4║5║6║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
G║3║2║2║2║2║2║3║4║5║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
T║4║3║3║3║3║3║3║3║4║
//! [alignment_matrix_format::unicode_block#score]
//! [alignment_matrix_format::unicode_block#trace]
 ║ε║A║A║C║C║G║G║T║T║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
ε║█║▖║▖║▖║▖║▖║▖║▖║▖║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
A║▝║▘║▌║▖║▖║▖║▖║▖║▖║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
C║▝║▝║▘║▘║▌║▖║▖║▖║▖║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
G║▝║▝║▀║▀║▘║▘║▌║▖║▖║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
T║▝║▝║▀║▀║▀║▀║▘║▘║▌║
//! [alignment_matrix_format::unicode_block#trace]
//! [alignment_matrix_format::unicode_braille#score]
 ║ε║A║A║C║C║G║G║T║T║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
ε║0║1║2║3║4║5║6║7║8║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
A║1║0║1║2║3║4║5║6║7║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
C║2║1║1║1║2║3║4║5║6║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
G║3║2║2║2║2║2║3║4║5║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
T║4║3║3║3║3║3║3║3║4║
//! [alignment_matrix_format::unicode_braille#score]
//! [alignment_matrix_format::unicode_braille#trace]
 ║ε║A║A║C║C║G║G║T║T║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
ε║⠀║⠄║⠄║⠄║⠄║⠄║⠄║⠄║⠄║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
A║⠈║⠁║⠅║⠄║⠄║⠄║⠄║⠄║⠄║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
C║⠈║⠈║⠁║⠁║⠅║⠄║⠄║⠄║⠄║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
G║⠈║⠈║⠉║⠉║⠁║⠁║⠅║⠄║⠄║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
T║⠈║⠈║⠉║⠉║⠉║⠉║⠁║⠁║⠅║
//! [alignment_matrix_format::unicode_braille#trace]
//! [alignment_matrix_format::unicode_arrows#score]
 ║ε║A║A║C║C║G║G║T║T║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
ε║0║1║2║3║4║5║6║7║8║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
A║1║0║1║2║3║4║5║6║7║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
C║2║1║1║1║2║3║4║5║6║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
G║3║2║2║2║2║2║3║4║5║
 ╬═╬═╬═╬═╬═╬═╬═╬═╬═╬
T║4║3║3║3║3║3║3║3║4║
//! [alignment_matrix_format::unicode_arrows#score]
//! [alignment_matrix_format::unicode_arrows#trace]
 ║ε ║A ║A ║C ║C ║G ║G ║T ║T ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
ε║↺ ║← ║← ║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
A║↑ ║↖ ║↖←║← ║← ║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
C║↑ ║↑ ║↖ ║↖ ║↖←║← ║← ║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
G║↑ ║↑ ║↖↑║↖↑║↖ ║↖ ║↖←║← ║← ║
 ╬══╬══╬══╬══╬══╬══╬══╬══╬══╬
T║↑ ║↑ ║↖↑║↖↑║↖↑║↖↑║↖ ║↖ ║↖←║
//! [alignment_matrix_format::unicode_arrows#trace]
